db/bigtable.go

```diff
@@ -642,324 +722,369 @@ func (bigtable *Bigtable) SaveProposals(blocks map[uint64]map[string]*types.Bloc
-				status = 0
-			}
+		vals := validators[i:upperBound]
 
-			if inclusionSlot > maxSlot {
-				maxSlot = inclusionSlot
+		g.Go(func() error {
+			select {
+			case <-gCtx.Done():
+				return gCtx.Err()
+			default:
 			}
+			ranges := bigtable.getValidatorsEpochRanges(vals, ATTESTATIONS_FAMILY, startEpoch, endEpoch)
+			err := bigtable.tableValidatorsHistory.ReadRows(ctx, ranges, func(r gcp_bigtable.Row) bool {
+				keySplit := strings.Split(r.Key(), ":")
 
-			validator, err := strconv.ParseUint(strings.TrimPrefix(ri.Column, ATTESTATIONS_FAMILY+":"), 10, 64)
-			if err != nil {
-				logger.Errorf("error parsing validator from column key %v: %v", ri.Column, err)
-				return false
-			}
+				validator, err := bigtable.validatorKeyToIndex(keySplit[1])
+				if err != nil {
+					logger.Errorf("error parsing validator from row key %v: %v", r.Key(), err)
+					return false
+				}
 
-			if attestationsMap[validator] == nil {
-				attestationsMap[validator] = make(map[uint64][]*types.ValidatorAttestation)
-			}
+				for _, ri := range r[ATTESTATIONS_FAMILY] {
+					attesterSlotString := strings.Replace(ri.Column, ATTESTATIONS_FAMILY+":", "", 1)
+					attesterSlot, err := strconv.ParseUint(attesterSlotString, 10, 64)
+					if err != nil {
+						logger.Errorf("error parsing slot from row key %v: %v", r.Key(), err)
+						return false
+					}
+					inclusionSlot := MAX_BLOCK_NUMBER - uint64(ri.Timestamp)/1000
 
-			if attestationsMap[validator][attesterSlot] == nil {
-				attestationsMap[validator][attesterSlot] = make([]*types.ValidatorAttestation, 0)
-			}
+					status := uint64(1)
+					if inclusionSlot == MAX_BLOCK_NUMBER {
+						inclusionSlot = 0
+						status = 0
+					}
 
-			attestationsMap[validator][attesterSlot] = append(attestationsMap[validator][attesterSlot], &types.ValidatorAttestation{
-				InclusionSlot: inclusionSlot,
-				Status:        status,
-			})
-		}
-		return true
-	}, gcp_bigtable.RowFilter(filter))
-	if err != nil {
+					resMux.Lock()
+					if attestationsMap[types.ValidatorIndex(validator)] == nil {
+						attestationsMap[types.ValidatorIndex(validator)] = make(map[types.Slot][]*types.ValidatorAttestation)
+					}
+
+					if attestationsMap[types.ValidatorIndex(validator)][types.Slot(attesterSlot)] == nil {
+						attestationsMap[types.ValidatorIndex(validator)][types.Slot(attesterSlot)] = make([]*types.ValidatorAttestation, 0)
+					}
+
+					attestationsMap[types.ValidatorIndex(validator)][types.Slot(attesterSlot)] = append(attestationsMap[types.ValidatorIndex(validator)][types.Slot(attesterSlot)], &types.ValidatorAttestation{
+						InclusionSlot: inclusionSlot,
+						Status:        status,
+					})
+					resMux.Unlock()
+
+				}
+				return true
+			}, gcp_bigtable.RowFilter(filter))
+
+			return err
+		})
+	}
+
+	if err := g.Wait(); err != nil {
 		return nil, err
 	}
 
 	// Find all missed and orphaned slots
 	slots := []uint64{}
+	maxSlot := (endEpoch + 1) * utils.Config.Chain.Config.SlotsPerEpoch
```

Comment:
**Question:** Shouldn't this be `maxSlot := ((endEpoch + 1) * utils.Config.Chain.Config.SlotsPerEpoch) - 1` as below we use `slot <= maxSlot`?